hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/staking/EndOfStakingPeriodCalculator.java

```diff
@@ -0,0 +1,412 @@
+
+import static com.hedera.node.app.service.mono.utils.Units.HBARS_TO_TINYBARS;
+import static com.hedera.node.app.service.token.impl.handlers.BaseCryptoHandler.asAccount;
+import static com.hedera.node.app.service.token.impl.staking.PeriodStakingUtils.calculateRewardSumHistory;
+import static com.hedera.node.app.service.token.impl.staking.PeriodStakingUtils.readableNonZeroHistory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.hedera.hapi.node.base.Fraction;
+import com.hedera.hapi.node.base.Timestamp;
+import com.hedera.hapi.node.state.token.NetworkStakingRewards;
+import com.hedera.hapi.node.state.token.StakingNodeInfo;
+import com.hedera.hapi.node.transaction.NodeStake;
+import com.hedera.hapi.node.transaction.NodeStakeUpdateTransactionBody;
+import com.hedera.hapi.node.transaction.TransactionBody;
+import com.hedera.node.app.service.token.ReadableAccountStore;
+import com.hedera.node.app.service.token.ReadableNetworkStakingRewardsStore;
+import com.hedera.node.app.service.token.impl.WritableNetworkStakingRewardsStore;
+import com.hedera.node.app.service.token.impl.WritableStakingInfoStore;
+import com.hedera.node.app.spi.numbers.HederaAccountNumbers;
+import com.hedera.node.app.spi.workflows.HandleContext;
+import com.hedera.node.app.spi.workflows.record.SingleTransactionRecordBuilder;
+import com.hedera.node.config.data.StakingConfig;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.math.MathContext;
+import java.math.RoundingMode;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+@Singleton
+public class EndOfStakingPeriodCalculator {
+    private static final Logger log = LogManager.getLogger(EndOfStakingPeriodCalculator.class);
+
+    // The exact choice of precision will not have a large effect on the per-hbar reward rate
+    private static final MathContext MATH_CONTEXT = new MathContext(8, RoundingMode.DOWN);
+
+    private final HederaAccountNumbers accountNumbers;
+    private final StakeInfoHelper stakeInfoHelper;
+
+    @Inject
+    public EndOfStakingPeriodCalculator(
+            @NonNull final HederaAccountNumbers accountNumbers, @NonNull final StakeInfoHelper stakeInfoHelper) {
+        this.accountNumbers = accountNumbers;
+        this.stakeInfoHelper = stakeInfoHelper;
+    }
+
+    /**
+     * Updates all (relevant) staking-related values for all nodes, as well as any network reward information,
+     * at the end of a staking period. This method must be invoked during handling of a transaction
+     *
+     * TODO: this method is still not called by the handle workflow
```

Comment:
We should probably create an issue and assign to hedera-base team or confirm if they already have issue for it

Reply:
There is an issue for it, and I've coded it already. it's issue #7721